source/puppet/4.8/env_isolation.md

```diff
@@ -0,0 +1,94 @@
+---
+layout: default
+title: "Isolating resources in multiple environments"
+canonical: "/puppet/latest/environment_isolation.html"
+description: "Generating metadata to isolate resources in environments in Puppet"
+---
+
+[code_mgr_env]: {{pe}}/code_mgr.html#environment-isolation-metadata
+
+
+If you have multiple environments, environment isolation prevents resource types from leaking between your various environments. To use environment isolation, you'll generate metadata files that Puppet can use in place of the normal Ruby resource type implementations.
+
+In open source Puppet, enable environment isolation with the `generate types` command. In Puppet Enterprise, environment isolation is automatically provided by [Code Manager][code_mgr_env]. Environment isolation is not supported for r10k with Puppet Enterprise.
+
+## Preventing resource types leaks in multiple environments
+
+If you use multiple environments with Puppet, you might encounter issues with multiple versions of the same resource type leaking between your various environments. 
+
+This problem occurs because Ruby resource type bindings are global in the Ruby runtime. Thus, the first loaded version of a Ruby resource type takes priority, and then subsequent requests to compile in other environments get that first loaded version.
+
+Environment isolation solves this issue by generating and using metadata that describes the resource type implementation, instead of using the Ruby resource type implementation when compiling catalogs. 
+
+> **Note**: Other environment isolation cases, such as external helper logic issues or varying versions of required gems, are not solved with the `generate type` command.
+
+### Agent logic is unchanged
+
+Catalog compilation results are exactly the same whether you use the metadata-based version of the resource type or the Ruby implementation. The agent acting on the catalog continues to use the Ruby-based implementation. Because the agent uses one, and only one, version of the logic, isolation is not a problem for the agent.
+
+## Enable environment isolation in Puppet
+
+With open source Puppet, enable environment isolation by running the `generate types` command:
+
+1. On the command line, run `puppet generate types --environment <envname>` for each of your environments.
+
+For example, to generate metadata for your production environment, run:
+
+`puppet generate types --environment production`
+
+Whenever you deploy a new version of Puppet, overwrite previously generated metadata with `puppet generate types --environment <envname> --force`.
+
+### Implement environment isolation with r10k
```

Comment:
You're not wrong, Walter ...

I guess my reaction when I saw this in the doc was "could this subsection be omitted altogether?"  But, now that I've typed that, I believe I more-or-less asked the same question when we last spoke about this in-person, so that definitely seems like 🔨  a 💀  🐴  territory.